### DISCUSSION TITLE
How to disambiguate many-many query?

### DISCUSSION THREAD

**Discussion by leesus** [supabase/supabase#1291] [Questions]
I am trying to query using a join table, but I have another table with foreign keys to my entities, and am struggling to find the magic incantation for my query to work.

I currently have 4 tables: `users`, `teams`, `team_users` (join table) and `notes` (has foreign keys to `users` and `teams`. The `notes` table is the one now causing the disambiguation error.

What I want to end up with is:
```
{
  user: {
    id: "blah",
    name: "blah",
    teams: [
      {
        id: "blah",
        name: "blah"
      }
    ]
  }
}
```

The following query is the closest I've managed to get:
```
db.from('users')
    .select(`
      *,
      teams:team_users!user_id(team:team_id(*))
    `)
    .eq('id', someuserid)
    .single();
```

Unfortunately, this results in (nested objects in teams):
```
{
  user: {
    ...
    teams: [
      {
        team: {
          id: "blah",
          name: "blah"
        }
      }
    ]
  }
}
```

Is there a query I can use to get my desired output, or is this the best I can get?

Thanks!

**Comment by leesus** [comment]
I dropped the `team_id` foreign key on `notes` table and instead added another join table for `team_notes` - this seemed to fix it and let me use a nicer GraphQL-like embedded resource query.

Open to hearing if there was/is a better way.

**Comment by steve-chavez** [comment]
Hey @leesus,

Can you try it like:

```js
db.from('users')
    .select(`
      *,
      teams!teams_users(*)
    `)
    .eq('id', someuserid)
    .single();
```

`teams!teams_users(*)` says that the M2M join should go through `teams_users`(and not through `notes`).

**Reply by alexstaroselsky** [reply]
@leesus Have you attempted to create a `user` and associate that user to a `team` (or similar) in a single operation? If so can you share a snippet of that code? Or are you having to execute a create in join table `team_users` to "connect" a team and a user? Thanks!

**Reply by leesus** [reply]
@alexstaroselsky I am doing it in multiple statements - I don't know if it's possible to do a transaction using the JS API, other than using a stored procedure. Only started with supabase yesterday!

**Reply by leesus** [reply]
@steve-chavez that's the first thing I tried, all that does is return the rows in the join table, not the entities themselves.

**Reply by alexstaroselsky** [reply]
@leesus Are you planning on writing policies based on group/team membership? As in, a user can only update a team or similar if they are part of that team. I imagine that would need some sort of join/exists that checks against team users? I ask because I’m looking for examples in the wild when dealing with many to many in supabase more effectively.

**Reply by steve-chavez** [reply]
@leesus Hm, the example I posted will return the `teams` rows not the join table rows(I've just tried with some sample tables). Perhaps you can double check?

Also, just in case, the docs about disambiguation are here: https://postgrest.org/en/v7.0.0/api.html#embedding-disambiguation

**Comment by hoangvu12** [comment]
Have you solved the problem? Maybe all we can do is using `.map` to get the data we need...

**Reply by steve-chavez** [reply]
Maybe try sharing a sample schema and request/response.

**Reply by hoangvu12** [reply]
```js
  const { data, error } = await supabase
    .from("anime")
    .select(
      "title->userPreferred, recommendations!original_id(recommendations:recommend_id(*))"
    )
    .limit(1)
    .single();
```

it returns 
```js
{
  userPreferred: 'Dragon Quest: Dai no Daibouken',
  recommendations: [
    { recommendations: [Object] },
    { recommendations: [Object] },
    { recommendations: [Object] }
  ]
}
```

The expected output is
```js
{
  userPreferred: 'Dragon Quest: Dai no Daibouken',
  recommendations: [
    [Object],
    [Object],
    [Object]
  ]
}
```

Is there anyway to do this?

**Reply by steve-chavez** [reply]
Hm, can you share the foreign keys or join tables between `anime` and `recommendations`?

In your example you're explicitly asking to double nest, it should be somewhat like:

```js
  const { data, error } = await supabase
    .from("anime")
    .select(
      "title->userPreferred, recommendations:recommend_id(*)"
    )
    .limit(1)
    .single();
```

**Reply by hoangvu12** [reply]
> Hm, can you share the foreign keys or join tables between `anime` and `recommendations`?
> 
> In your example you're explicitly asking to double nest, it should be somewhat like:
> 
> ```js
>   const { data, error } = await supabase
>     .from("anime")
>     .select(
>       "title->userPreferred, recommendations:recommend_id(*)"
>     )
>     .limit(1)
>     .single();
> ```

> Anime
- id
- ani_id (different id that's not generated by supabase)

> Recommendations
- id
- original_id --> anime.ani_id (The anime that recommend_id is recommended to)
- recommend_id --> anime.ạni_id (The recommended anime)

Sorry if the structure confuses you because I'm totally new to sql.

**Reply by steve-chavez** [reply]
I see. You want a many-to-many to the same table(`anime`) using a different foreign key in the junction(`recommendations`). Unfortunately this is not possible at the moment - it's a [pending enhancement](https://github.com/PostgREST/postgrest/issues/1907). 

Right now you could do it in SQL and then call it through [rpc](https://supabase.io/docs/reference/javascript/rpc), as suggested on https://github.com/supabase/supabase/discussions/1969#discussioncomment-876358.

**Reply by hoangvu12** [reply]
> I see. You want a many-to-many to the same table(`anime`) using a different foreign key in the junction(`recommendations`). Unfortunately this is not possible at the moment - it's a [pending enhancement](https://github.com/PostgREST/postgrest/issues/1907).
> 
> Right now you could do it in SQL and then call it through [rpc](https://supabase.io/docs/reference/javascript/rpc), as suggested on [#1969 (reply in thread)](https://github.com/supabase/supabase/discussions/1969#discussioncomment-876358).

Thank you, I'll learn how to write sql language, ehe

**Reply by hoangvu12** [reply]
> I see. You want a many-to-many to the same table(`anime`) using a different foreign key in the junction(`recommendations`). Unfortunately this is not possible at the moment - it's a [pending enhancement](https://github.com/PostgREST/postgrest/issues/1907).
> 
> Right now you could do it in SQL and then call it through [rpc](https://supabase.io/docs/reference/javascript/rpc), as suggested on [#1969 (reply in thread)](https://github.com/supabase/supabase/discussions/1969#discussioncomment-876358).

By the way, Is there any way to write a custom method to supabase ? Like this.

```js
  const { data, error } = await supabase
    .from("anime")
    .select(
      "title->userPreferred, recommendations!original_id(recommendations:recommend_id(*))"
    )
    .map() // custom method that i made just to get the correct data.
    .limit(1)
    .single();
```
